### PR TITLE
Simplify our Poco mappers

### PR DESCRIPTION
**Pull request by zpqrtbnk**
See priv.284 - adresses startup time improvement.

This PR simplifies how we register our Poco mappers, going from expressions (eg `x => x.Id`) to names (eg `nameof(Thing.Id)`) - because processing all the expressions when starting was time and memory consuming (including some potential JIT etc). We don't need the expressions, really.

Review = code review + all tests (should be green) + some site testing, but that should be rather simple.

### CODE REVIEW

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs [hunk 1]
```diff
@@ -10,22 +8,38 @@ namespace Umbraco.Core.Persistence.Mappers
...
+            if (!_maps.TryGetValue(GetType(), out var mapperMaps))
+                throw new InvalidOperationException($"No maps defined for mapper {GetType().FullName}.");
+            if (!mapperMaps.TryGetValue(propertyName, out var mappedName))
+                throw new InvalidOperationException($"No map defined by mapper {GetType().FullName} for property {propertyName}.");
+            return mappedName;
         }
 
-        protected abstract void BuildMap();
-
+        // fixme kill
```
Comment (Shazwazza) [inline src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: Should you kill this then?

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs [hunk 2]
```diff
@@ -40,8 +54,36 @@ internal string Map(ISqlSyntaxProvider sqlSyntax, string propertyName, bool thro
...
+
+            var columnName = columnAttribute.Name;
+            var columnMap = _sqlSyntax.GetQuotedTableName(tableName) + "." + _sqlSyntax.GetQuotedColumnName(columnName);
+
+            var mapperMaps = _maps.GetOrAdd(GetType(), type => new ConcurrentDictionary<string, string>());
+            mapperMaps[sourceName] = columnMap;
+        }
 
-        internal DtoMapModel ResolveMapping<TSource, TDestination>(Expression<Func<TSource, object>> sourceMember, Expression<Func<TDestination, object>> destinationMember)
+        // fixme kill
```
Comment (Shazwazza) [inline src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: should be killed then?

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs (hunk 1 above)
Comment (zpqrtbnk) [reply src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: I've been lazy and forgot to check for fixmes before pushing - clear now.

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs [hunk 3]
```diff
@@ -10,60 +8,47 @@ namespace Umbraco.Core.Persistence.Mappers
...
         {
-            BuildMap();
+            _sqlSyntax = sqlContext.SqlSyntax;
+            _maps = maps;
         }
 
-        protected abstract void BuildMap();
-
-        internal string Map(ISqlSyntaxProvider sqlSyntax, string propertyName, bool throws = false)
+        internal string Map(string propertyName, bool throws = false)
```
Comment (bergmania) [inline src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: The `throws` variable is not used anymore

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs [hunk 4]
```diff
@@ -10,60 +8,47 @@ namespace Umbraco.Core.Persistence.Mappers
...
         }
 
-        internal void CacheMap<TSource, TDestination>(Expression<Func<TSource, object>> sourceMember, Expression<Func<TDestination, object>> destinationMember)
+        //protected void DefineMap<TSource, TTarget>(string sourceName, Expression<Func<TTarget, object>> targetMember)
+        protected void DefineMap<TSource, TTarget>(string sourceName, string targetName)
         {
-            var property = ResolveMapping(sourceMember, destinationMember);
-            PropertyInfoCache.AddOrUpdate(property.SourcePropertyName, property, (x, y) => property);
-        }
+            var sourceType = typeof(TSource);
```
Comment (bergmania) [inline src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: The `sourceType` variable is not used anymore

**Comment by bergmania** [comment]
I tested this, and I can't find any errors anymore.

File: src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs [hunk 5]
```diff
@@ -10,60 +8,74 @@ namespace Umbraco.Core.Persistence.Mappers
...
+        private readonly Lazy<ISqlContext> _sqlContext;
+        private readonly object _definedLock = new object();
+        private readonly ConcurrentDictionary<Type, ConcurrentDictionary<string, string>> _maps;
 
-        internal abstract ConcurrentDictionary<string, DtoMapModel> PropertyInfoCache { get; }
+        private ISqlSyntaxProvider _sqlSyntax;
+        private bool _defined;
 
-        private void Build()
+        protected BaseMapper(Lazy<ISqlContext> sqlContext, ConcurrentDictionary<Type, ConcurrentDictionary<string, string>> maps)
```
Comment (Shazwazza) [inline src/Umbraco.Core/Persistence/Mappers/BaseMapper.cs]: These are essentially breaking changes, though i don't suspect any public users are using these APIs, in either case, will need to mark this as a breaking change.

**Pull request merged**